pkg/segment/segexecution.go

```diff
@@ -157,52 +157,99 @@ func HelperQueryArithmeticAndLogical(queryOps []structs.QueryArithmetic, resMap
+				// So, if we want to determine whether there are elements with the same labels in another metric, we need to appropriately modify the group ID.
+				rGroupID := ""
+				labelStr := ""
+				if len(lGroupID) >= len(resultLHS.MetricName) {
+					labelStr = lGroupID[len(resultLHS.MetricName):]
+					rGroupID = resultRHS.MetricName + labelStr
+				}
+
+				if queryOp.Operation == utils.LetOr || queryOp.Operation == utils.LetUnless {
+					labelStrSet[labelStr] = struct{}{}
+				}
+
+				if _, ok := resultRHS.Results[rGroupID]; !ok && queryOp.Operation != utils.LetOr && queryOp.Operation != utils.LetUnless {
 					continue
 				} //Entries for which no matching entry in the right-hand vector are dropped
-				finalResult[groupID] = make(map[uint32]float64)
+				finalResult[lGroupID] = make(map[uint32]float64)
 				for timestamp, valueLHS := range tsLHS {
-					valueRHS := resultRHS.Results[groupID][timestamp]
+					valueRHS := resultRHS.Results[rGroupID][timestamp]
 					switch queryOp.Operation {
 					case utils.LetAdd:
-						finalResult[groupID][timestamp] = valueLHS + valueRHS
+						finalResult[lGroupID][timestamp] = valueLHS + valueRHS
 					case utils.LetDivide:
 						if valueRHS == 0 {
 							continue
 						}
-						finalResult[groupID][timestamp] = valueLHS / valueRHS
+						finalResult[lGroupID][timestamp] = valueLHS / valueRHS
 					case utils.LetMultiply:
-						finalResult[groupID][timestamp] = valueLHS * valueRHS
+						finalResult[lGroupID][timestamp] = valueLHS * valueRHS
 					case utils.LetSubtract:
-						finalResult[groupID][timestamp] = valueLHS - valueRHS
+						finalResult[lGroupID][timestamp] = valueLHS - valueRHS
 					case utils.LetGreaterThan:
 						if valueLHS > valueRHS {
-							finalResult[groupID][timestamp] = valueLHS
+							finalResult[lGroupID][timestamp] = valueLHS
 						}
 					case utils.LetGreaterThanOrEqualTo:
 						if valueLHS >= valueRHS {
-							finalResult[groupID][timestamp] = valueLHS
+							finalResult[lGroupID][timestamp] = valueLHS
 						}
 					case utils.LetLessThan:
 						if valueLHS < valueRHS {
-							finalResult[groupID][timestamp] = valueLHS
+							finalResult[lGroupID][timestamp] = valueLHS
 						}
 					case utils.LetLessThanOrEqualTo:
 						if valueLHS <= valueRHS {
-							finalResult[groupID][timestamp] = valueLHS
+							finalResult[lGroupID][timestamp] = valueLHS
 						}
 					case utils.LetEquals:
 						if valueLHS == valueRHS {
-							finalResult[groupID][timestamp] = valueLHS
+							finalResult[lGroupID][timestamp] = valueLHS
 						}
 					case utils.LetNotEquals:
 						if valueLHS != valueRHS {
-							finalResult[groupID][timestamp] = valueLHS
+							finalResult[lGroupID][timestamp] = valueLHS
 						}
+					case utils.LetAnd:
+						finalResult[lGroupID][timestamp] = valueLHS
+					case utils.LetOr:
+						finalResult[lGroupID][timestamp] = valueLHS
+					case utils.LetUnless:
+						finalResult[lGroupID][timestamp] = valueLHS
+					}
```

Comment:
I was confused for a while with this part, until I found that:
1. The `And` operation doesn't have to check the RHS here because from lines 175-176, we only get here if the left and right sides have the same labels or the operation is `Or` or `Unless`
2. We do the extra handling for `Or` and `Unless` a bit lower on lines 226-251 

I think it would be good to either rearrange some code or at least add some comments about the flows for these operations, to make it more clear